sklearn/tree/tests/test_tree.py

```diff
@@ -1950,6 +1966,30 @@ def test_apply_path_readonly_all_trees(name):
     check_apply_path_readonly(name)
 
 
+@pytest.mark.parametrize("criterion", ["mse", "friedman_mse", "poisson"])
+@pytest.mark.parametrize("Tree", REG_TREES.values())
+def test_balance_property(criterion, Tree):
+    # Test that sum(y_pred)=sum(y_true) on training set.
+    # This works if the mean is predicted (should even be true for each leaf).
+    # MAE predicts the median and is therefore excluded from this test.
+
+    # Choose a training set with non-negative targets (for poisson)
+    X, y = diabetes.data, diabetes.target
+    reg = Tree(criterion=criterion)
+    reg.fit(X, y)
+    assert np.sum(reg.predict(X)) == pytest.approx(np.sum(y))
+
+
+def test_poisson_zero_nodes():
+    # Test that sum(y)=0 and therefore y_pred=0 is forbidden on nodes.
+    X = [[0, 0], [0, 1], [0, 2], [0, 3],
+         [1, 0], [1, 2], [1, 2], [1, 3]]
+    y = [0, 0, 0, 0, 1, 2, 3, 4]
+    reg = DecisionTreeRegressor(criterion="poisson", random_state=1)
+    reg.fit(X, y)
+    assert np.all(reg.predict(X) > 0)
```

Comment:
ping on this one ;)

Reply:
I added a further dataset where **this test fails**:bomb:. As the impurity improvement returns -INFINITY (same as initialized value), it should not allow those splits. I'm investigating the splitter code where this might happen.
Interestingly, the impurity value for those nodes that predict zero is `np.nan`.

Reply:
> Interestingly, the impurity value for those nodes that predict zero is np.nan.

Is this issue specific to this PR or an general issue with the tree splitting?

Reply:
The `np.nan`was a consequence of a split for Poisson that was not allowed with only y=0 in the node. This led to `y_pred=0` which produced Poisson loss = np.nan.
This is all gone with commit 82e35867c2aa01e3c2f944c100564a799dc01655 "BUG check for y_mean <= 0 in poisson_loss".